CloudyLauncher.java

```diff
@@ -0,0 +1,497 @@
+                setToken(responseToken);
+
+                setFeedback("User recognised.");
+
+                initialiseGamePanel();
+            }   
+
+        } catch (IOException e) {
+            setFeedback("Check server.");
+
+        }
+    }
+
+    private void attemptUserRegistration(String username, String password, String email, String firstName, String lastName) {
+
+        try {
+            URL url = new URL(baseurl + "/users/");        
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Accept", "application/json");
+            String queryData = String.format("username=%s&password=%s&email=%s&first_name=%s&last_name=%s", username, password, email, firstName, lastName);
+
+            connection.setDoOutput(true);
+            DataOutputStream writer = new DataOutputStream(connection.getOutputStream());
+
+            writer.writeBytes(queryData);
+            writer.flush();
+            writer.close();
+
+            if (connection.getResponseCode() == HttpURLConnection.HTTP_BAD_REQUEST) {   
+                String errorMessage = "";
+
+                BufferedReader errorReader = new BufferedReader(new InputStreamReader(connection.getErrorStream()));
+                JSONObject errorResponse = new JSONObject(errorReader.readLine());
+
+                if (errorResponse.has("username")) {
+                    errorMessage = errorMessage + "\nUsername: " + errorResponse.getJSONArray("username").getString(0);
+                }
+
+                if (errorResponse.has("password")) {
+                    errorMessage = errorMessage + "\nPassword: " + errorResponse.getJSONArray("password").getString(0); 
+                }
+
+                if (errorResponse.has("email")) {
+                    errorMessage = errorMessage + "\nEmail: " + errorResponse.getJSONArray("email").getString(0); 
+                } 
+
+                if (errorResponse.has("first_name")) {
+                    errorMessage = errorMessage + "\nFirst Name: " + errorResponse.getJSONArray("first_name").getString(0); 
+                }
+
+                if (errorResponse.has("last_name")) {
+                    errorMessage = errorMessage + "\nLast Name: " + errorResponse.getJSONArray("last_name").getString(0); 
+                }
+
+                setFeedback(errorMessage);
+
+            } else {
+
+                setFeedback("User successfully registered.");
+            }                
+
+        } catch (IOException e) {
+            setFeedback("Check server.");
+        }
+    }
+
+    /**
+     * Returns the Game object containing the given information.
+     * 
+     * @param gameObj  JSONObject containing game information
+     * @return         Game object with information 
+     */    
+    private Game getGameFromJson(JSONObject gameObj) {
```

Comment:
Not sure but, is it possible to separate the logic (join game, get controller id, process game list) and the UI (show panel, buttons, etc) to different files? I think this file will grow longer as we support more functionalities (buy game, quit game, etc) so it'd be easier to maintain if they are separated. If it's not possible, maybe move it into sections (part it) in this file?